pydantic/_internal/_typing_extra.py

```diff
@@ -281,18 +286,17 @@ def ForwardRefWrapper(arg: Any, is_argument: bool = True, *, is_class: bool = Fa
 
         Implemented as EAFP with memory.
         """
-        global fr_has_is_class
-
-        if not fr_has_is_class:
-            return typing.ForwardRef(arg, is_argument)
-
+        global make_forward_ref
         try:
-            return typing.ForwardRef(arg, is_argument, is_class=is_class)
+            res = typing.ForwardRef(arg, is_argument, is_class=is_class)  # type: ignore
+            make_forward_ref = typing.ForwardRef  # type: ignore
+            return res
         except TypeError:
-            fr_has_is_class = False
             return typing.ForwardRef(arg, is_argument)
 
-    ForwardRef = ForwardRefWrapper  # noqa F811
```

Comment:
Pyright was not happy with this. It's quite strict as to what is a type and not, and assigning a function to a type is not okay with it. This change makes it explicit that we have a wrapper and not a type. It also simplifies the memory logic by just overriding itself on the first call if it can.